strawberryfields/apps/graph/similarity.py

```diff
@@ -0,0 +1,125 @@
+r"""
+Graph similarity
+================
+
+**Module name:** :mod:`strawberryfields.apps.graph.similarity`
+
+.. currentmodule:: strawberryfields.apps.graph.similarity
+
+Functionality for calculating feature vectors of graphs using GBS.
+
+Summary
+-------
+
+.. autosummary::
+    sample_to_orbit
+    sample_to_event
+    orbits
+
+Code details
+^^^^^^^^^^^^
+"""
+from typing import Generator
+
+
+def sample_to_orbit(sample: list) -> list:
+    """Provides the orbit corresponding to a given sample.
+
+    Orbits are simply a sorting of samples in non-increasing order with the zeros at the end
+    removed.
+
+    Example usage:
+
+    >>> sample = [1, 2, 0, 0, 1, 1, 0, 3]
+    >>> sample_to_orbit(sample)
+    [3, 2, 1, 1, 1]
+
+    Args:
+        sample (list[int]): a sample from GBS
+
+    Returns:
+        list[int]: the orbit of the sample
+    """
+    sorted_sample = sorted(sample, reverse=True)
+    return [c for c in sorted_sample if c != 0]
+
+
+def sample_to_event(sample: list, max_count_per_mode: int) -> list:
+    """Provides the event corresponding to a given sample.
+
+    Example usage:
+
+    >>> sample = [1, 2, 0, 0, 1, 1, 0, 3]
+    >>> sample_to_event(sample, 4)
+    8
+
+    >>> sample_to_event(sample, 2)
+    None
+
+    Args:
+        sample (list[int]): a sample from GBS
+        max_count_per_mode (int): the maximum number of photons counted in any given mode for a
+            sample to be categorized as an event. Samples with counts exceeding this value are
+            attributed the event ``None``.
+
+    Returns:
+        list[int]: the orbit of the sample
+    """
+    if max(sample) <= max_count_per_mode:
+        return sum(sample)
+
+    return None
+
+
+def orbits(photon_number: int) -> Generator[list, None, None]:
+    """Generate all the possible orbits for a given photon number.
+
+    Provides a generator over the integer partitions of ``photon_number``.
+    Code derived from `website <http://jeromekelleher.net/generating-integer-partitions.html>`__
+    of Jerome Kelleher's, which is based upon an algorithm from Ref. :cite:`kelleher2009generating`.
+
+    Example usage:
+
+    >>> o = orbits(5)
+    >>> list(o)
+    [[1, 1, 1, 1, 1], [2, 1, 1, 1], [3, 1, 1], [2, 2, 1], [4, 1], [3, 2], [5]]
+
+    Args:
+        photon_number (int): number of photons to generate orbits from
+
+    Returns:
+        Generator[list[int]]: orbits with total photon number adding up to ``photon_number``
+    """
+    a = [0 for _ in range(photon_number + 1)]
+    k = 1
+    y = photon_number - 1
+    while k != 0:
+        x = a[k - 1] + 1
+        k -= 1
+        while 2 * x <= y:
+            a[k] = x
+            y -= x
+            k += 1
+        l = k + 1
+        while x <= y:
+            a[k] = x
+            a[l] = y
+            yield sorted(a[: k + 2], reverse=True)
+            x += 1
+            y -= 1
+        a[k] = x + y
+        y = x + y - 1
+        yield sorted(a[: k + 1], reverse=True)

```

Comment:
Learnt about generators thanks to this! 
Using generators is probably overkill for the usage we expect for this function, but I'm in favour keeping as is. The code is stable and employing state-of-the-art methods.